packages/engine/errors/7061.md

```diff
@@ -0,0 +1,34 @@
+---
+original: 'A mapped type may not declare properties or methods.'
+excerpt: 'You can only create a mapped type using a type alias.'
+---
+
+Are you accidentally creating a mapped type with an interface?
```

Comment:
So a mapped type is an object type whose properties are dynamic or generic which can only be created with `type` aliases:
```ts
type CamelCaseName = {
  [P in CamelCaseNameKeys]: string;
};
```
As Matt showed in previous comment playground, the error implies that a mapped object cannot contain both static properties/methods and dynamic ones:
```ts
type CamelCaseName = {
  [P in CamelCaseNameKeys]: string; // << dynamic
  Foo: string // << static
};
```
# So why does it error on interfaces?
Well, interfaces can **only** contain static properties (even if the interface is empty) whereas type aliases can contain either of both static or dynamic.

# Exceptions - Advanced use of interfaces:
I'll contradict myself here 😅 . You _can_ create an interface with a mapped type, as long as it's not generic:
```ts
type InternalCamelCaseName = {
  [P in CamelCaseNameKeys]: string;
};
interface CamelCaseName extends InternalCamelCaseName {}
```
This is an advanced way to convert mapped types to interfaces so I don't think it should actually be mentioned here.

Another exception is:
```ts
interface Foo {
  [key: string]: any;
}
```
Because it's not really dynamic or generic but rather means any property is `string`. This is allowed as long as keys are not literal types or generic. However, don't worry about this last case since it displays a totally different error.

# How to explain this in short?

So, I don't know 😅 . The answer is something along the lines:
> A mapped type cannot contain both: static and dynamic properties

The description could explain this case with this example:
```ts
type Keys = "Foo" | "Bar"
type MyMappedType = {
  [Key in Keys]: string; // << dynamic properties
  Baz: string            // << static properties
};
```
The less relevant part imho is about the interface, but you could mention something like:
> Interfaces can only contain static properties (even if empty), that's why attempting to add dynamic properties is disallowed (remember: you can't have both in the same type).